src/main/java/org/prebid/server/bidder/huaweiads/HuaweiAdsBidder.java

```diff
@@ -0,0 +1,1267 @@
+        return app.build();
+    }
+
+    private String getFinalPackageName(String bundleName) {
+        for (PkgNameConvert convert : packageNameConverter) {
+            if (convert.getConvertedPkgName().isEmpty()) {
+                continue;
+            }
+
+            for (String name : convert.getExceptionPkgNames()) {
+                if (name.equals(bundleName)) {
+                    return bundleName;
+                }
+            }
+
+            for (String name : convert.getUnconvertedPkgNames()) {
+                if (name.equals(bundleName) || name.equals("*")) {
+                    return convert.getConvertedPkgName();
+                }
+            }
+
+            for (String keyword : convert.getUnconvertedPkgNameKeyWords()) {
+                if (bundleName.indexOf(keyword) > 0) {
+                    return convert.getConvertedPkgName();
+                }
+            }
+
+            for (String prefix : convert.getUnconvertedPkgNamePrefixs()) {
+                if (bundleName.startsWith(prefix)) {
+                    return convert.getConvertedPkgName();
+                }
+            }
+        }
+
+        return bundleName;
+    }
+
+    private Device getDeviceInfo(BidRequest openRTBRequest) {
+        final Device.DeviceBuilder deviceBuilder = Device.builder();
+        Optional.ofNullable(openRTBRequest.getDevice())
+                .ifPresent(deviceData -> {
+                    deviceBuilder.type(deviceData.getDevicetype());
+                    deviceBuilder.useragent(deviceData.getUa());
+                    deviceBuilder.os(deviceData.getOs());
+                    deviceBuilder.version(deviceData.getOsv());
+                    deviceBuilder.maker(deviceData.getMake());
+                    deviceBuilder.model(StringUtils.isEmpty(deviceData.getModel())
+                            ? HuaweiAdsConstants.DEFAULT_MODEL_NAME : deviceData.getModel());
+                    deviceBuilder.height(deviceData.getH());
+                    deviceBuilder.width(deviceData.getW());
+                    deviceBuilder.language(deviceData.getLanguage());
+                    deviceBuilder.pxratio(deviceData.getPxratio());
+                    String countryCode = getCountryCode(openRTBRequest);
+                    deviceBuilder.belongCountry(countryCode);
+                    deviceBuilder.localeCountry(countryCode);
+                    deviceBuilder.ip(deviceData.getIp());
+                    deviceBuilder.gaid(deviceData.getIfa());
+                });
+
+        return getDeviceIdFromUserExt(deviceBuilder.build(), openRTBRequest);
+    }
+
+    private String getCountryCode(BidRequest openRTBRequest) {
+        return Stream.of(openRTBRequest)
+                .filter(request -> request.getDevice() != null
+                        && request.getDevice().getGeo() != null
+                        && !request.getDevice().getGeo().getCountry().isEmpty())
+                .findFirst()
```

Comment:
Why do you use stream for single element??? 
Also, there are too many code duplications in this method, this should be entirely rewritten and common parts have to be extracted to separate methods.